File: benchmarks/pandas/bench_frequencies.py
"""Benchmark: frequencies — to_offset and infer_freq.
Tests parsing frequency strings and inferring frequency from date arrays.
"""
import json
import time
import pandas as pd

WARMUP = 5
ITERATIONS = 50

FREQ_STRINGS = ["D", "h", "min", "s", "ME", "MS", "YE", "YS", "W", "3ME", "2h", "QE", "QS"]

# Build a regularly-spaced daily date index for infer_freq
daily_index = pd.date_range("2020-01-01", periods=365, freq="D")


def run():
    for freq in FREQ_STRINGS:
        pd.tseries.frequencies.to_offset(freq)
    pd.infer_freq(daily_index)


for _ in range(WARMUP):
    run()

start = time.perf_counter()
for _ in range(ITERATIONS):
    run()
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "frequencies",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_frequencies.ts
/**
 * Benchmark: frequencies — toOffset and inferFreq.
 * Tests parsing frequency strings and inferring frequency from date arrays.
 */
import { toOffset, inferFreq } from "../../src/index.js";

const WARMUP = 5;
const ITERATIONS = 50;

const FREQ_STRINGS = ["D", "h", "min", "s", "ME", "MS", "YE", "YS", "W", "3ME", "2h", "QE", "QS"];

// Build a regularly-spaced daily date array for inferFreq
const BASE = new Date(Date.UTC(2020, 0, 1));
const DAILY_DATES = Array.from({ length: 365 }, (_, i) => new Date(BASE.getTime() + i * 86_400_000));

function run() {
  for (const freq of FREQ_STRINGS) {
    toOffset(freq);
  }
  inferFreq(DAILY_DATES);
}

for (let i = 0; i < WARMUP; i++) {
  run();
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  run();
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "frequencies",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
